lustre/ws.lua

```diff
@@ -0,0 +1,351 @@
+                    self.socket:close()
+                    return
+                end
+                goto continue
+            end
+
+            --should we close because we have been waiting to long for a ping?
+            -- we might not need to do this...it doesn't seem like it was prioritized
+            -- with a test case in autobahn
+            if pending_pong then
+                frames_since_last_ping = frames_since_last_ping + 1
+                if frames_since_last_ping > self.config._max_frames_without_pong then
+                    frames_since_last_ping = 0
+                    local err = self:close(CloseCode.policy(), "no pong after ping")
+                end
+            end
+
+
+            --handle fragmentation
+            if frame.header.opcode.sub == 'text' then
+                msg_type = "text"
+                if multiframe_message then --we expected a continuation message
+                    self:close(CloseCode.protocol(), "expected "..msg_type.."continuation frame")
+                    goto continue
+                end
+                if not frame:is_final() then
+                    multiframe_message = true
+                end
+            elseif frame.header.opcode.sub == 'binary' then
+                msg_type = "binary"
+                if multiframe_message then
+                    self:close(CloseCode.protocol(), "expected "..msg_type.."continuation frame")
+                    goto continue
+                end
+                if not frame:is_final() then
+                    multiframe_message = true
+                end
+            elseif frame.header.opcode.sub   == 'continue' and not multiframe_message then
+                self:close(CloseCode.protocol(), "unexpected continue frame")
+                goto continue
+            end
+            --aggregate payloads
+            if not frame:is_final() then --todo and we are continuing something
+                received_bytes = received_bytes + frame:payload_len()
+                --Dont build up a message payload that is bigger than max message size
+                if received_bytes <= self.config.max_message_size then
+                    table.insert(partial_frames, frame.payload)
+                end
+                goto continue
+            else
+                multiframe_message = false
+            end
+
+
+            --coalesce frame payloads into single message payload
+            local full_payload = frame.payload
+            if next(partial_frames) then
+                table.insert(partial_frames, frame.payload)
+                full_payload = table.concat(partial_frames)
+                partial_frames = {}
+            end
+            if self.message_cb then
+                self.message_cb(Message.new(msg_type, full_payload))
+            end
+        elseif recv[1] == self._rx then
+            self._rx:recieve()
+            --TODO @FreeMasen what events were you thinking would be injected into 
```

Comment:
I was expecting this task would handle all of the socket communications. So the two `send` methods above would send on this channel and this task would then handle the `socket:send`, most likely with a retry on timeout, and sending in multiple parts if needed.

Reply:
I think I have made this asychronous so that th erun loop handles the socket operations. However, the retry on timeout and the testing of the fragmentation has yet to be completed.